lib/Sema/TypeChecker.h

```diff
@@ -892,8 +890,11 @@ class TypeChecker final : public LazyResolver {
   /// member.
   /// \param useArchetypes Whether to use context archetypes for outer generic
   /// parameters if the class is nested inside a generic function.
+  /// \param stage The type resolution stage for type validation,
+  /// defaults to Interface.
   static Type substMemberTypeWithBase(ModuleDecl *module, TypeDecl *member,
-                                      Type baseTy, bool useArchetypes = true);
+                                      Type baseTy, bool useArchetypes = true,
+                                      TypeResolutionStage stage = TypeResolutionStage::Interface);
```

Comment:
Maybe if you pass in a TypeResolution instead of a TypeResolutionStage, you can combine the useArchetypes and state parameters into one; useArchetypes then just becomes resolution.usesArchetypes().